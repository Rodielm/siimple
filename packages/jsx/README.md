# @siimple/jsx (Experimental)

A simple JSX transform function to create React Elements that supports the `css` prop.

> **Note**: this is an experimental package. API may change at any time.

## Installation

Install this package using **npm**:

```bash
 npm install --save @siimple/jsx
```

## Usage

This transform allows you to provide a custom CSS styles object using a `css` prop:

```js
/** @jsxImportSource @siimple/jsx */

const App = () => (
    <div
        align="center"
        css={{
            color: "cadetblue",
            fontFamily: "Roboto, sans-serif",
            padding: "4rem",
            "&:hover": {
                cursor: "pointer",
                color: "navy",
            },
        }}
    >
        <strong>Hello world</strong>
    </div>
);
```

You can use `ThemeProvider` from `@siimple/theme-provider` for using your own theme:

```js
/** @jsxImportSource @siimple/jsx */
import ThemeProvider from "@siimple/theme-provider";

const theme = {
    colors: {
        primary: "cadetblue",
    },
};

const App = () => (
    <ThemeProvider theme={theme}>
        <div css={{color: "primary"}}>
            <strong>Hello world</strong>
        </div>
    </ThemeProvider>
);
```

Remember that you should tell Babel to use `@siimple/jsx` as the [automatic runtime import](https://babeljs.io/docs/en/babel-plugin-transform-react-jsx#customizing-the-automatic-runtime-import):

```js
/** @jsxImportSource @siimple/jsx */

const App = () => (
    <div css={{color: "blue"}}>
        <strong>Hello world</strong>
    </div>
);
```

You can also use it as the [classic runtime](https://babeljs.io/docs/en/babel-plugin-transform-react-jsx#customizing-the-classic-runtime-import):

```js
/** @jsx jsx */
import {jsx} from "@siimple/jsx";

const App = () => (
    <div css={{color: "blue"}}>
        <strong>Hello world</strong>
    </div>
);
```

## API

### jsx(type, props, ...children)

The JSX transform function that converts the `props.css` property to a className.

### styled(type, css)

Generate a styled React component with the specified type and CSS.

```jsx
import {styled} from "@siimple/jsx";

const ButtonLink = styled("a", {
    color: "primary",
    display: "block",
    padding: "2rem",
    textDecoration: "none",
    "&:hover": {
        cursor: "pointer",
    },
});

const App = () => (
    <div align="center">
        <ButtonLink href="/login">
            Login
        </ButtonLink>
    </div>
);
```

## License

[MIT License](https://github.com/jmjuanes/siimple/blob/main/LICENSE).
