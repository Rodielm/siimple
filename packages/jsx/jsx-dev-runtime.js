import {jsxDEV as ReactJsxDEV} from "react/jsx-dev-runtime";
import {SiimpleElement, parseProps} from "./index.js";

export {Fragment} from "react/jsx-dev-runtime";

// https://github.com/facebook/react/blob/9ff738f53f228ea4343886a1e719ec2f1533d300/packages/react/src/jsx/ReactJSXElementValidator.js#L297-L304
export const jsxDEV = (type, props, key, isStaticChildren, source, self) => {
    if (!props?.css || typeof props?.css !== "object") {
        return ReactJsxDEV(type, props, key, isStaticChildren, source, self);
    }
    return ReactJsxDEV(SiimpleElement, parseProps(type, props), key, isStaticChildren, source, self);
};
