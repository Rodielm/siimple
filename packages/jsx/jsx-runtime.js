import {jsx as ReactJsx, jsxs as ReactJsxs} from "react/jsx-runtime";
import {SiimpleElement, parseProps} from "./index.js";

export {Fragment} from "react/jsx-runtime";

export const jsx = (type, props, key) => {
    if (!props?.css || typeof props?.css !== "object") {
        return ReactJsx(type, props, key);
    }
    return ReactJsx(SiimpleElement, parseProps(props, type), key);
};

export const jsxs = (type, props, key) => {
    if (!props?.css || typeof props?.css !== "object") {
        return ReactJsxs(type, props, key);
    }
    return ReactJsxs(SiimpleElement, parseProps(props, type), key);
};
