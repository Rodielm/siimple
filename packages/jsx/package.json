{
    "name": "@siimple/jsx",
    "version": "0.3.0-beta",
    "description": "A simple JSX transform function to support the css prop.",
    "author": "Josemi Juanes <josemi@siimple.xyz>",
    "type": "module",
    "license": "MIT",
    "repository": "https://github.com/jmjuanes/siimple",
    "bugs": "https://github.com/jmjuanes/siimple/issues",
    "homepage": "https://www.siimple.xyz",
    "main": "index.js",
    "keywords": [
        "siimple",
        "react",
        "jsx"
    ],
    "dependencies": {
        "@siimple/styled": "^0.2.3",
        "@siimple/theme-provider": "^0.3.0-beta",
        "react": "^17.0.2"
    }
}
