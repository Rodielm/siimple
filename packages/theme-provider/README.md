# @siimple/theme-provider (Experimental)

A wrapper component for providing access to your theme in all React components.

## Installation

Install this package using **npm**:

```bash
 npm install --save @siimple/theme-provider
```

## Usage

```jsx
import React from "react";
import {ThemeProvider, useTheme} from "@siimple/theme-provider";

const defaultTheme = {
    colors: {
        primary: "coral",
    },
};

const MyText = props => {
    const theme = useTheme();
    
    return (
        <span style={{color: theme.colors.primary}}>
            {props.children}
        </span>
    );
};

export default App = () => (
    <ThemeProvider theme={defaultTheme}>
        <MyText>Hello world</MyText>
    </ThemeProvider>
);

```


## API

### ThemeProvider

A simple React component that accepts a **siimple theme object** as a prop and pass it top-down using [Context](https://reactjs.org/docs/context.html).

```jsx
import {ThemeProvider} from "@siimple/theme-provider";

export default App = () => (
    <ThemeProvider theme={theme}>
        ...
    </ThemeProvider>
);
```


### useTheme

A React [Hook](https://reactjs.org/docs/hooks-intro.html) that returns the current theme provided to `<ThemeProvider>`.

```jsx
import {useTheme} from "@siimple/theme-provider";

export const Text = props => {
    const theme = useTheme();

    return (
        <span style={{color: theme.colors.primary}}>
            {props.children}
        </span>
    );
};
```


## License

[MIT License](https://github.com/jmjuanes/siimple/blob/main/LICENSE).
