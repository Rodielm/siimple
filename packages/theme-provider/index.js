import React from "react";

// Theme context
export const ThemeContext = React.createContext({});

// Use theme hook
// Returns: the current theme configuration object
export const useTheme = () => {
    return React.useContext(ThemeContext);
};

// Theme provider component
export const ThemeProvider = props => {
    let theme = useTheme();
    if (props.theme && typeof props.theme === "object" && props.theme !== theme) {
        theme = props.theme;
    }
    return React.createElement(ThemeContext.Provider, {value: theme}, props.children);
};

// HOC that wraps a component and adds the current theme as a prop
export const withTheme = Component => {
    const name = Component.displayName || Component.name || "Component";
    const ThemedComponent = React.forwardRef((props, ref) => {
        const theme = useTheme();

        return React.createElement(Component, {theme, ref, ...props}, props.children);
    });
    ThemedComponent.displayName = `withTheme(${name})`;
    return ThemedComponent;
};
