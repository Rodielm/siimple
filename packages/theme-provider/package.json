{
    "name": "@siimple/theme-provider",
    "version": "0.3.0-beta",
    "description": "A wrapper component for providing access to your theme in all React Components.",
    "author": "Josemi Juanes <josemi@siimple.xyz>",
    "type": "module",
    "license": "MIT",
    "repository": "https://github.com/jmjuanes/siimple",
    "bugs": "https://github.com/jmjuanes/siimple/issues",
    "homepage": "https://www.siimple.xyz",
    "main": "index.js",
    "keywords": [
        "siimple",
        "react",
        "theme",
        "theme-provider"
    ],
    "dependencies": {
        "react": "^17.0.2"
    }
}
