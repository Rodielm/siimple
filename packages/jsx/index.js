import React from "react";
import {
    classNames,
    createCache,
    registerCss,
} from "@siimple/styled";
import {useTheme} from "@siimple/theme-provider";

// Global variables
const TYPE_PROP = "__siimple_type_prop__";

// Cache Context
const CacheContext = React.createContext({
    current: createCache({
      key: "siimple-react",
    }),
});

// Element wrapper for jsx
const SiimpleElement = React.forwardRef((props, ref) => {
    const theme = useTheme();
    const cache = React.useContext(CacheContext);

    if (typeof props.css !== "object") {
        throw new Error(`[siimple:jsx] Invalid 'css' prop: expected 'Object' but got '${typeof props.css}'`);
    }

    // Generate styles (the css prop always exists)
    const className = registerCss(props.css, theme, cache.current, false, false);

    // Generate element props
    const newProps = {};
    Object.keys(props).forEach(key => {
        if (key !== TYPE_PROP && key !== "css" && key !== "children") {
            newProps[key] = props[key];
        }
    });

    // Assign new reference and classname
    newProps.ref = ref;
    newProps.className = classNames(className, props.className);

    // Return createElement
    return React.createElement(props[TYPE_PROP], newProps, props.children);
});

// jsx wrapper
export const jsx = (...args) => {
    const type = args[0];
    const props = args[1];
    
    // No props or props.css provided --> just call React.createElement
    if (!props?.css) {
        return React.createElement(...args);
    }

    // Generate new props for our wrapper element
    const newProps = {
        ...(props || {}),
        [TYPE_PROP]: type || "div",
    };
    return React.createElement(SiimpleElement, newProps, ...args.slice(2));
};
