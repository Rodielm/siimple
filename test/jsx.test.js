/**
 * @jest-environment jsdom
 */

import renderer from "react-test-renderer";
import {jsx} from "@siimple/jsx";
import {ThemeProvider} from "@siimple/theme-provider";

const selector = `style[data-siimple="siimple-react"]`;

describe("jsx", () => {
    it("should return a React.createElement if no 'css' prop is provided", () => {
        const component = renderer.create(
            jsx("div", {}, "Hello world"),
        );

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("should style the element", () => {
        const component = renderer.create(
            jsx("div", {
                css: {
                    color: "red",
                },
            }, "Hello world in red"),
        );
        const div = component.root.findByType("div");

        expect(div).toBeDefined();
        expect(div.props.className).toEqual(expect.stringContaining("sii-"));
        expect(document.querySelector(selector).innerHTML).toEqual(expect.stringContaining("color:red"));
    });

    it("should style the element using the theme", () => {
        const theme = {
            colors: {
                primary: "black",
            },
        };
        const component = renderer.create(
            jsx(ThemeProvider, {theme},
                jsx("div", {css: {color: "primary"}}, "Hello world"),
            ),
        );
        const div = component.root.findByType("div");

        expect(div).toBeDefined();
        expect(div.props.className).toEqual(expect.stringContaining("sii-"));
        expect(document.querySelector(selector).innerHTML).toEqual(expect.stringContaining("color:black"));
    });
});
