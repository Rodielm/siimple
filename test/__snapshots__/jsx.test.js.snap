// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsx should return a React.createElement if no 'css' prop is provided 1`] = `
<div>
  Hello world
</div>
`;
